### ISSUE TITLE
Aggressive Memory Leak

### ISSUE THREAD

**Issue by RHavar** [tj/co#180]
I've been able to narrow down an aggressive memory leak I'm experiencing to this:

Run with:  node --harmony --expose-gc filename.js

``` javascript
var co = require('co');

function* sleep() {
    return new Promise(function(resolve) {
        setTimeout(resolve, 1);
    });
};

co(function*() {

    for(var i = 0; true; ++i) {
        yield sleep();

        if (i % 10000 === 0) {
            global.gc();
            console.log(process.memoryUsage());
        }

    }

}).then(function() {
    console.log('finished')
}, function(err) {
    console.log('caught error: ', err.stack);
});
```

Sample output:

```
{ rss: 17420288, heapTotal: 9620736, heapUsed: 3590768 }
{ rss: 44822528, heapTotal: 49288192, heapUsed: 12972200 }
{ rss: 70955008, heapTotal: 58575616, heapUsed: 21688912 }
{ rss: 80048128, heapTotal: 66831104, heapUsed: 30531560 }
{ rss: 89157632, heapTotal: 76118528, heapUsed: 39490184 }
{ rss: 98275328, heapTotal: 85405952, heapUsed: 48445040 }
{ rss: 107368448, heapTotal: 93661440, heapUsed: 57410024 }
{ rss: 116477952, heapTotal: 102948864, heapUsed: 66365712 }
{ rss: 125591552, heapTotal: 112236288, heapUsed: 75330040 }
{ rss: 134684672, heapTotal: 120491776, heapUsed: 84285144 }
{ rss: 143798272, heapTotal: 129779200, heapUsed: 93250072 }
{ rss: 152907776, heapTotal: 139066624, heapUsed: 102205152 }
{ rss: 162000896, heapTotal: 147322112, heapUsed: 111170352 }
{ rss: 171114496, heapTotal: 156609536, heapUsed: 120125032 }
```

**Comment by hax** [comment]
@jonathanong 
V8 bug? 
But I tested Chrome 39, it seems already be fixed.

**Comment by dead-horse** [comment]
@hax you should read this: https://github.com/promises-aplus/promises-spec/issues/179

As they discussed, it's a bug in spec, [bluebird](https://github.com/petkaantonov/bluebird/commit/8c1edaf0a77a6d46a7527b2873e456d3ff62fab8) and [then/promise](https://github.com/then/promise/pull/67) already fixed this in their implementations, but these patches don't follow the spec.

**Comment by jonathanong** [comment]
if someone wants to test the old version in iojs/master... :D

**Comment by fengmk2** [comment]
I use co@~4.10 still got memory leak if we remove `global.gc();`.

Test scripts https://gist.github.com/fengmk2/8fdbd896fee458efe0b4

**Comment by hax** [comment]
@fengmk2 V8 only gc if short of memory.

**Comment by yanickrochon** [comment]
@fengmk2 a memory leak is when calling the garbage collector does not work as expected. Garbage collecting is an heavy process that may slow the program down, so V8 (as other languages using gc) calls it automatically when short of memory, or if idle (i.e. waiting for IO or something), or when the engine notices an increase in orphan objects (i.e. ready to be gc'd). In the sample you provided, you see that memory is, indeed, garbage collected at some point. You have to compare memory usage at startup, and monitor peaks; these peaks should not increase excessively during a burn-in test.

**Comment by petkaantonov** [comment]
> I use co@~4.10 still got memory leak if we remove `global.gc();`.

If calling `global.gc()` frees memory then it's not a memory leak.

**Comment by fengmk2** [comment]
My problem, sorry for disturb all.

**Comment by undoZen** [comment]
@fengmk2 I tried run your demo (https://gist.github.com/fengmk2/8fdbd896fee458efe0b4) in a memory limited container (iojs/latest), gc triggered when limit reached.

**Comment by dead-horse** [comment]
anybody can take a look at https://github.com/cnpm/mirrors/pull/41?

**Comment by fengmk2** [comment]
@undoZen the memory limit value is 100mb, 200mb or 1gb?

**Comment by undoZen** [comment]
@fengmk2 I tried 20m 22m 25m all take over 2~4m over specified limit but not gone further.

**Comment by hax** [comment]
@dead-horse After reading the issue (https://github.com/promises-aplus/promises-spec/issues/179 ) and writing a promise implementation my self, I believe that this is NOT the spec bug, just missing an optional optimization for edge case. And I believe it's possible to optimize the promise implementation to avoid this issue while not break the right behavior about the order. I will publish my new promise implementaion to prove it soon.

**Comment by undoZen** [comment]
@hax hope you will keep your promise ... to release your promise implementation soon.

**Comment by petkaantonov** [comment]
@hax [it was already proven](https://github.com/promises-aplus/promises-spec/issues/179#issuecomment-68316012) however I don't see it as an optional optimization (memory leak is a bug) or an edge case (the pattern is the clearest one to perform for loop equivalent).

**Comment by hax** [comment]
@petkaantonov It seems still have some issues. The main problem is, V8 native impl (Chrome and Safari 8) and Firefox native impl show very diff behaviors. Currently Firefox may be more close to the ES6 spec as my understanding. I will post a testcase soon.

**Comment by hax** [comment]
I recently found that this issue is more complex than I thought, to preserve the order behavior of current ES6 draft strictly (which most implementations fail, only Firefox implementation and my-promise is ok), it's impossible to eliminate "leak" completely, though it can be reduced to leaking only one small object (as a pointer to denote the position) for each nest promise.
So, it may be a ES6 spec "bug" if you don't treat this as edge case. But it may be not a Promise/A+ spec bug since Promise/A+ spec seems not define the order. I opened an issue https://github.com/promises-aplus/promises-spec/issues/183 for more discuss.
